worker/sshserver/worker.go

```diff
@@ -0,0 +1,141 @@
+	NewServerWorker func(ServerWorkerConfig, bool) (*ServerWorker, error)
+	Logger          Logger
+}
+
+// Validate validates the workers configuration is as expected.
+func (c ServerWrapperWorkerConfig) Validate() error {
+	// TODO(ale8k): Once the PR for implementing configuration is merged, check
+	// the host key is populated here only. For now, this check is better than nothing.
+	if c.StateInfo == (controller.StateServingInfo{}) {
+		return errors.NotValidf("StateInfo is required")
+	}
+	if c.StatePool == nil {
+		return errors.NotValidf("StatePool is required")
+	}
+	if c.NewServerWorker == nil {
+		return errors.NotValidf("NewSSHServer is required")
+	}
+	if c.Logger == nil {
+		return errors.NotValidf("Logger is required")
+	}
+	return nil
+}
+
+// serverWrapperWorker is a worker that runs an ssh server worker.
+type serverWrapperWorker struct {
+	// catacomb holds the catacomb responsible for running the server worker
+	// and configuration watcher workers.
+	catacomb catacomb.Catacomb
+
+	// config holds the configuration required by the server wrapper worker.
+	config ServerWrapperWorkerConfig
+}
+
+// NewServerWrapperWorker returns a new worker that runs an ssh server worker internally.
+// This worker will listen for changes in the controller configuration and restart the
+// server worker when the port or max concurrent connections changes.
+func NewServerWrapperWorker(config ServerWrapperWorkerConfig) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &serverWrapperWorker{
+		config: config,
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+// Kill implements worker.Worker.
+func (ssw *serverWrapperWorker) Kill() {
+	ssw.catacomb.Kill(nil)
+}
+
+// Wait implements worker.Worker.
+func (ssw *serverWrapperWorker) Wait() error {
+	return ssw.catacomb.Wait()
+}
+
+// loop is the main loop of the server wrapper worker. It starts the server worker
+// and listens for changes in the controller configuration.
+func (ssw *serverWrapperWorker) loop() error {
+	srv, err := ssw.config.NewServerWorker(ServerWorkerConfig{
+		Logger: ssw.config.Logger,
+	}, true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := ssw.catacomb.Add(srv); err != nil {
+		return errors.Trace(err)
+	}
+
+	systemState, err := ssw.config.StatePool.SystemState()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	controllerConfig := systemState.WatchControllerConfig()
+	if err := ssw.catacomb.Add(controllerConfig); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-ssw.catacomb.Dying():
+			return ssw.catacomb.ErrDying()
+		case <-controllerConfig.Changes():
+			// TODO(ale8k): Once the configuration PR is merged, get the max conns & port
+			// from controller config. Get the HostKey from server info, and feed them through
+			// to NewServerWorker.
```

Comment:
i suppose we only need to restart the ssh server worker if one of the relevant config values changes, not for every controller config change, right?

Reply:
We can't know what has changed, so it's for every change. In 4.0 they have a means to get the diff.

Reply:
You can do a request to the controller config and see if the key has changed, it does mean you have to store the old value here.

This can be done in another PR.